### ISSUE TITLE
Confusion between loading table in psql session vs python db driver - memory not shared?

### ISSUE THREAD

**Issue by pavitrakumar78** [knizhnik/imcs#60]
Hi,

I've been trying to write some scrips using python's psycopg2 db driver and I'm not able to figure out how the table load works and whether the loaded table is shared between all db connections or not.

For example:
- assume I just restarted the server and nothing is in the memory
1. I open up a python shell and execute:
```
import psycopg2

PGSQL_USER = 'postgres'
PGSQL_PASS = '<>'
PGSQL_PORT = 5434
PGSQL_DBNAME = 'sample_db'

connection = psycopg2.connect(user=PGSQL_USER,
                              password=PGSQL_PASS,
                              host="127.0.0.1",
                              port=PGSQL_PORT,
                              database=PGSQL_DBNAME)
cursor = connection.cursor()
qry = "select Quotes_load();"
cursor.execute(qry)
qry_res = cursor.fetchall()
```
This completes execution successfully and running the following queries:
```
qry = "select cs_used_memory()/(1024*1024);"
cursor.execute(qry)
qry_res = cursor.fetchall()

qry = "select Quotes_is_loaded();"
cursor.execute(qry)
qry_res = cursor.fetchall()
```
Gives output as 1350 and True which is consistent with the table size I've loaded and I now know that the table has been loaded.

2. Now if I open up a psql session in another cmd (while python shell opened in step 1 is still active) and execute:
```
select us_eod_is_loaded();
```
The query gets stuck - i.e, I see nothing in the logs and it's still in running phase.

Why does this happen? 
~~-I should be able to load it once via db connection made in python and execute imcs commands on both psql and from within python shell where I created a db session right?~~ new observation below
~~-Are there any caveats with executing IMCS queries with an ORM like sqlalchemy because of this issue?~~ new observation below

Edit:
new observation: 
As soon as I manually closed the connection (`connection.close()`) in python shell, the queries in psql started working.

-Was the python db connection blocking psql session's access to imcs(?). Is this expected behavior? 
-Is there a definite possibility that the same table will be loaded twice when executed from different sessions like the experiment above? (I observed this few times, but can't reproduce this exactly)
-If this is a limitation, are there any ways of executing individual queries in parallel using different sessions using an ORM or does that not work because of the way the loaded table works in imcs?

**Comment by knizhnik** [comment]
Please do 
cursor.close()

**Comment by pavitrakumar78** [comment]
> Please do
> cursor.close()

Thank you! I tried this and it works.

Another issue is - when executing ~50-60 queries one after another (all more or less the same date range and symbols) on a table size of ~1.5 GB, the memory usage kept climbing from the initial 1.5GB (from after `table_load()`) all the way to >11GB! 

All the queries return only a single number and I do close the cursor after calling `fetchall`. If I'm performing a lot of window aggregations but on the same set of data, shouldn't the computed/extra data (other than the loaded table) be discarded after each query since we close the cursor? 

At first, I thought this was ORM/db connector specific but then just to be sure I also tried executing the queries normally one by one and the same happened. Is this due to query caching? and is it possible to remove this and reduce memory consumption **buildup**? In my application, it's very uncommon that the same query gets executed twice in a short span.

**Comment by knizhnik** [comment]
Can you send me a query for which you can get this backend memory bloat?
If you are using _cs_project, try to disable project cacheing:
`imcs.project_caching=false`

**Comment by pavitrakumar78** [comment]
> Can you send me a query for which you can get this backend memory bloat?
> If you are using _cs_project, try to disable project cacheing:
> `imcs.project_caching=false`

I've mentioned both the SQL query and the IMCS equivalent in my previous comment. I'm trying to calculate the ADX indicator. Here are the links to queries with proper table/col names:
IMCS version: https://pastebin.com/ezUhkME4
PG12 version: https://pastebin.com/AUdRJMkj
(I've tried to stay consistent with the `Quote` naming in the imcs docs)

I've verified the output of IMCS version by looking at a single stock and checking with a trading platform.
But this query is too big to manually debug and it'd take you much lesser time to just reconstruct the query from the pseudocode below. Few other experiments I've tried is to try and reduce the amount of computation by nesting the query but I can't see any improvement in the query runtime. Here's an example of a nested linear regression calculation vs untested/single linear regression query in imcs: https://pastebin.com/341BsN2G - I thought nesting would reduce the computation required but didn't see any change in the final query in the explain analyze output. 

The Source of the ADX formula:
```
input length = 14;
input averageType = AverageType.WILDERS;

# high[1] = accessing previous high
def hiDiff = high - high[1];
def loDiff = low[1] - low;

def plusDM = if hiDiff > loDiff and hiDiff > 0 then hiDiff else 0;
def minusDM =  if loDiff > hiDiff and loDiff > 0 then loDiff else 0;

def ATR = MovingAverage(averageType, TrueRange(high, close, low), length);
plot "DI+" = 100 * MovingAverage(averageType, plusDM, length) / ATR;
plot "DI-" = 100 * MovingAverage(averageType, minusDM, length) / ATR;

def DX = if ("DI+" + "DI-" > 0) then 100 * AbsValue("DI+" - "DI-") / ("DI+" + "DI-") else 0;
plot ADX = MovingAverage(averageType, DX, length);
```

**Comment by knizhnik** [comment]
Sorry, but I was not able to reproduce the problem (constantly growing backend's memory).
I have used your query https://pastebin.com/ezUhkME4 with some fixes (Quote -> Quote_get, close_price -> Close)
and repeated it several times. During query execution virtual memory of backend grows to 6 Gb, but then reduced to 4Gb.
And it is quite stable and doesn't change after execution of query several times.

**Comment by pavitrakumar78** [comment]
> Sorry, but I was not able to reproduce the problem (constantly growing backend's memory).
> I have used your query https://pastebin.com/ezUhkME4 with some fixes (Quote -> Quote_get, close_price -> Close)
> and repeated it several times. During query execution virtual memory of backend grows to 6 Gb, but then reduced to 4Gb.
> And it is quite stable and doesn't change after execution of query several times.

Can you please try executing this script: https://pastebin.com/JcW9hXZh
Requirements: `psycopg2`, `psutil`, `sqlalchemy`
Here's my output (I already have table loaded):
```
Total memory: 16322.68 MB
Available memory: 12271.52 MB
[(0,)]
[(1326,)]
[(True,)]
table loaded in 0.45 secs
Available memory after loading table: 12222.63 MB
Available memory after query execution [1]: 11531.97 MB
Available memory after query execution [2]: 737.68 MB
Available memory after query execution [3]: 657.99 MB
```

What the script does: executes the same query (adx) 3 times. Based on what you have mentioned, I expected the memory usage to be constant after every query run but it's not - available memory decreases after the first run and stays there as long as the python shell is open. You can try executing the python script as a whole or try to copy and paste it into your python shell - you'll be able to see that the memory does not come down even after execution of the same query 3rd time. 

For the execution of the query, I have added ways to execute using psycopg2 and via sqlalchemy. Both give the same result of memory being the same after the execution of the query as long as the python terminal is open.  

Based on your previous suggestion, I have also used `cursor.close()` (in psycopg) and similarly, `session.close()` (for sqlalchemy) after every query. 

Am I missing something here? i.e, not handling imcs session properly?

**Comment by knizhnik** [comment]
This is what I get at my system:

```
knizhnik@knizhnik:~/db$ python3 adx.py 
Total memory: 15917.34 MB
Available memory: 10615.30 MB
[(0,)]
[(238,)]
[(True,)]
table loaded in 0.00 secs
Available memory after loading table: 10613.58 MB
Available memory after query execution [1]: 10609.84 MB
Available memory after query execution [2]: 10617.96 MB
Available memory after query execution [3]: 10609.84 MB
```

**Comment by pavitrakumar78** [comment]
Thank you for testing and reporting the results you got.

This is very weird... I just tried another fresh installation of imcs and got the same result as what I have mentioned in my previous comment (memory reducing after 2nd run). 

I will try to reset my postgres config and even try on a fresh VPS online and report back on what happens. 

Do you have any ideas on why I get this behavior?

**Comment by knizhnik** [comment]
How are you building and running IMCS? With default parameters? Or with some custom options? First of all: are you using disk mode?

**Comment by pavitrakumar78** [comment]
> How are you building and running IMCS? 
Fresh install : 
`git clone https://github.com/knizhnik/imcs.git`
`cd imcs`
`sudo ./install.sh`

add imcs to `shared_preload_libraries` in pg conf and restart server

> With default parameters?  Or with some custom options? 

Only default.

> First of all: are you using disk mode?

No.

**Comment by pavitrakumar78** [comment]
Update on my Ubuntu dedicated VPS experiment: It works! Just like the output you had - the memory is constant.  I don't know why this issue comes up only in my local dev machine's WSL instances... 

If you have anything in mind or anything you want me to try about this issue, let me know. Otherwise, I'll just create a workout (using new db connection etc.,) for local dev or just stick to a cheap VPS for dev work. 

Other than this, do you have any recommended settings for pg conf file or imcs conf to keep memory consumption low for complex queries like ADX (trade-off with time is fine)?
